packages/xstate-react/src/useMachine.ts

```diff
@@ -0,0 +1,246 @@
+): [
+  State<TContext, TEvent, any, TTypestate>,
+  Interpreter<TContext, any, TEvent, TTypestate>['send'],
+  Interpreter<TContext, any, TEvent, TTypestate>
+] {
+  if (process.env.NODE_ENV !== 'production') {
+    const [initialMachine] = useState(machine);
+
+    if (machine !== initialMachine) {
+      console.warn(
+        'Machine given to `useMachine` has changed between renders. This is not supported and might lead to unexpected results.\n' +
+          'Please make sure that you pass the same Machine as argument each time.'
+      );
+    }
+  }
+
+  const {
+    context,
+    guards,
+    actions,
+    activities,
+    services,
+    delays,
+    state: rehydratedState,
+    ...interpreterOptions
+  } = options;
+
+  const service = useConstant(() => {
+    const machineConfig = {
+      context,
+      guards,
+      actions,
+      activities,
+      services,
+      delays
+    };
+
+    const createdMachine = machine.withConfig(machineConfig, {
+      ...machine.context,
+      ...context
+    } as TContext);
+
+    return interpret(createdMachine, interpreterOptions).start(
+      rehydratedState ? State.create(rehydratedState) : undefined
+    );
+  });
+
+  const [state, setState] = useState(service.state);
+
+  const effectActionsRef = useRef<
+    Array<[ReactActionObject<TContext, TEvent>, State<TContext, TEvent>]>
+  >([]);
+  const layoutEffectActionsRef = useRef<
+    Array<[ReactActionObject<TContext, TEvent>, State<TContext, TEvent>]>
+  >([]);
+
+  useLayoutEffect(() => {
+    service.onTransition((currentState) => {
+      if (currentState.changed) {
+        setState(currentState);
+      }
+
+      if (currentState.actions.length) {
+        const reactEffectActions = currentState.actions.filter(
+          (action): action is ReactActionObject<TContext, TEvent> => {
+            return (
+              typeof action.exec === 'function' &&
+              '__effect' in (action as ReactActionObject<TContext, TEvent>).exec
+            );
+          }
+        );
+
+        const [effectActions, layoutEffectActions] = partition(
+          reactEffectActions,
+          (action): action is ReactActionObject<TContext, TEvent> => {
+            return action.exec.__effect === ReactEffectType.Effect;
+          }
+        );
+
+        effectActionsRef.current.push(
+          ...effectActions.map<ActionStateTuple<TContext, TEvent>>(
+            (effectAction) => [effectAction, currentState]
+          )
+        );
+
+        layoutEffectActionsRef.current.push(
+          ...layoutEffectActions.map<ActionStateTuple<TContext, TEvent>>(
+            (layoutEffectAction) => [layoutEffectAction, currentState]
+          )
+        );
+      }
+    });
+
+    // if service.state has not changed React should just bail out from this update
+    setState(service.state);
+
+    return () => {
+      service.stop();
+    };
+  }, []);
+
+  useLayoutEffect(() => {
+    while (layoutEffectActionsRef.current.length) {
```

Comment:
> If the execution of a queued effect leads to queuing an effect then when should it happen?

This will never happen (immediately), as the only way an action can possibly do this is by sending an event to the service, which might queue new actions only after these actions have been flushed.

We can add a test for this if I'm wrong, though.

Reply:
While doing `asEffect(() => send(FOO))` is definitely a weird thing to do, so it's kinda unrealistic, it can be somewhere deep down in the stack. But flattening more complex case which could happen in a real app to this simplistic `asEffect(() => send(FOO))` leads me to believe that an effect action could get in fact be executed in the middle of this flushing.

After all, this is being flushed whenever React decides to do so, so XState is not aware of this, it is happening outside of its internal scheduler which means that first `send` will execute immediately which in turn will cause `onTransition` to be called synchronously (still while flushing) and this might lead to a new effect action to be queued at the end of the currently flushed items list.

Reply:
Can we add a test for this? I don't believe this potential edge-case is a showstopper.

Reply:
sure, gonna add a test for this